_includes/articles/payment-batching.md

```diff
@@ -0,0 +1,256 @@
+bumping.  In addition, the more unconfirmed children a transaction has,
+the more RBF fee bumping will cost as you'll have to pay for both the
+increased feerate of your transaction as well as for all the potential
+fees lost to miners when they remove any child transactions in order
+to accept your replacement.
+
+Note that these problems are not unique to batched payments---independent
+payments can have the same problem.  However, if an independent payment
+can't be fee bumped because the independent receiver spent their output,
+only that user is affected.  But if a single receiver of a batched
+payment spends their output to the point where fee bumping becomes
+impossible, all the other receivers of that transaction are also affected.
+
+As of Bitcoin Core 0.18 (April 2019), the limits are[^package-limits] that a
+group of related unconfirmed transactions may not exceed 101,000 vbytes
+in size, have more than 25 unconfirmed ancestors, or have more than 25
+descendants.  This size limit restricts batches to a maximum size of
+about 3,000 outputs and the descendant limit is easily reached if just a
+tiny percentage of those receiving a large batch respend their confirmed
+outputs.  It's also easy for any of the receivers to deliberately create
+transactions that reach one of these limits and prevent fee bumping if
+they know that you're relying on that capability.
+
+## Implementation
+
+Payment batching is extremely easy using certain existing wallet
+implementations, such as using Bitcoin Core's [sendmany][] RPC.  Check your software
+documentation for a function that allows you to send multiple payments.
+
+```bash
+bitcoin-cli sendmany "" '{
+  "bc1q5c2d2ue7x38hcw2ugk5q7y4ae7nw4r6vxcptu8": 0.1,
+  "bc1qztjzd7hpf2xmngr7zkgkxsvdqcv2jpyfgwgtsv": 0.2,
+  "bc1qsul9emtnz0kks939egx2ssa6xnjpsvgwq9chrw": 0.3
+}'
+```
+
+<!-- for max standard tx size: src/policy/policy.h:static const unsigned int MAX_STANDARD_TX_WEIGHT = 400000; -->
+
+If using your own implementation, you are probably already creating
+transactions with two outputs in most cases (a payment output and a
+change output), so it should be easy to add
+support for additional outputs.  The only notable consideration is that
+Bitcoin Core nodes (and most other nodes) will refuse to accept or relay
+transactions over 100,000 vbytes, so you should not attempt to send
+batched payments larger than this.
+
+## Recommendations summary
+
+1. Try to create systems where your users and customers don't expect
+   their payments immediately but are willing to wait for some time
+   (the longer the better).
+
+2. Use low-feerate consolidations to keep some large inputs available
+   for spending.
+
+3. Within each time window, send all payments together in the same
+   transaction.  For example, create an hourly [cronjob][] that sends all pending payments.
+   Ideally, your prior consolidations should allow the
+   transaction to contain only a single input.
```

Comment:
What's the point of the second and third limits?  The second only makes sense to me if the goal is to ensure your transaction is under the 100,000 vbyte limit (in which case, maybe make the limit "a maximum transaction size").  The third only makes sense to me to ensure the value is under what you have in the wallet being used (e.g. the hot wallet); in which case, there should probably already be some withdrawal limit mechanism in place.

Reply:
Output count: the text argued earlier that batching has diminishing returns. Even in the optimal case of only a single input, e.g. in a transaction with 100 outputs, the cost per output is 31.78 vbytes. For a transaction with one input and 200 outputs, the cost only goes down only 1% to 31.39 vbytes.
On the other hand, even when customers may wait multiple confirmations before being fully confident that they have been paid (although in my experience most people are happy after one or two confirmations even amounts that I would consider sizeable), end-customers do tend to be antsy about seeing the service initiate the payment as soon as possible. Either because they just want to wrap up the open thread in their mind, or because they are looking to use the withdrawn amount elsewhere. Surely there is some expectation management involved there, but in my experience, services aim to keep the time until sending very brief because of the reduction in support burden it brings (some literally batch multiple times per minute, just to show the customers the transactions as quickly as possible). In a way the output count limit bounds the total customer wait time, still allowing the service to wait a bit longer for a bigger batch when there is little traffic in the middle of the night, but pushing out the transaction quicker in the rush hour when there is ten customers waiting after a minute already.

Reply:
Total value: The text noted that once the value of the summed outputs exceeds the value of the largest available inputs, the addition of further inputs reduces the efficiency of batching. In practice, many send-and-receive hot wallets tend to mill through their bigger deposits rather quickly, because at higher fees they can be used to build the most space efficient transactions. 
Some services consolidate consequently and have great stock in large UTXO, but especially in periods of longer congestion, consolidation transactions tend to be delayed.
Once the wallet gets into the bulk of their UTXO at e.g. ~0.02–0.07 BTC, they'll combine multiple inputs to fund their batches anyway. In that case, it may make sense to have a dynamic limit for transaction values at e.g. the value of 20 times a representative bulk UTXO. Keeping a few of the largest UTXO in stock is pertinent to be able to process the occasional huge withdrawal.

Reply:
So, in sum, I'd argue that delimiting transaction size leads to chains of unconfirmed transactions exceeding the ancestor limits less often, enables quicker turnaround for end-users, and has only marginal efficiency loss versus the biggest possible transactions.

Reply:
> Output count: the text argued earlier that batching has diminishing returns. [...] On the other hand, [...] end-customers do tend to be antsy about seeing the service initiate the payment as soon as possible.

I agree about users tending to be antsy---I've been that user wondering, *oh, crap, did I enter the wrong address?  Did I just blow a month's income?*---but I guess the Spock part of my brain says that, if you're willing to make the people initiating withdrawls during slow periods wait up to *x* amount of time, you ought to be consistent and make everyone wait up to that amount of time.  Varying the maximum amount of waiting time by current demand seems inconsistent.  Also, it means your previous sends had less time to get confirmed and so you need a larger UTXO pool.  (E.g. a wallet that only sends once per hour should work 99.99% of the time with just two UTXOs, which makes implementing consolidation easier.)

> Keeping a few of the largest UTXO in stock is pertinent to be able to process the occasional huge withdrawal.

That makes sense: the earlier you send, the earlier you confirm (all other things being equal).  Though maybe the rule should be tailored to change value rather than tx value.

(An alternative, but non-technical, reason to send high-total-value txes quickly is perhaps just to please you platform's whales.)

Reply:
> to ensure the value is under what you have in the wallet being used

So, basically, a good batching implementation definitely needs to track the value of the pending queue anyway. I'd just argue that it is probably better to have it lower than 100% of the hot wallet. If dynamic maybe rather 5% of the hot wallet.

> (E.g. a wallet that only sends once per hour should work 99.99% of the time with just two UTXOs, which makes implementing consolidation easier.)

I don't think that's necessarily true. If there is a mempool congestion, all funds could be in flight within an hour for a wallet with two UTXOs. Is that assuming that the spender uses unconfirmed inputs and automatically CPFPs unconfirmed preceding transactions? I've had a lot of fun this spring with thrifty services that mixed low feerate and high feerate sends as well as permitting any unconfirmed change unspents. There are a bunch of traps there that are fairly obvious when one understands how the mempool and UTXOs work, but there is a reason that certain wallet service providers avoid unconfirmed inputs by default in input selection. ;)

I am wondering whether we are just generally thinking about wallets of very different sizes. When I read the term high-frequency spender, I'm thinking at least 1,000+ payments per day, but it sounds to me that you might have been thinking about a different scale.

Reply:
I think the second point (batch size) is good and should be added. If the fee savings from no batching to batching 100 payments is ~75% and the savings from batching 100 payments to 200 payments is an additional ~1%, then it makes sense for the service provider to broadcast a batch as soon as the marginal gains from a larger batch are no longer worth it. It seems like this would make it less likely to suffer issues around tx pinning and fee bumping too.

I'm less convinced of the third point (batch value) since I think it's difficult to give universal advice. The best practice seems very dependent on lots of factors specific to the wallet/service.